### ISSUE TITLE
Invoice Refund Option

### ISSUE THREAD

**Issue by pavlenex** [btcpayserver/btcpayserver#890]
**Is your feature request related to a problem? Please describe.**
When a merchant wants to issue a refund to the customer (especially if the merchant is not auto-converting fiat, but hodling) issuing a refund is very difficult, especially if the service/product is priced in fiat.

Use case: Order total is $100, however one item is defective and merchants needs to issue a refund. The price for the item is $25.

To issue a refund currently, the merchant (if he's hodling) needs to refund $25 at the rate of the time of placing the order. This means that the merchant needs to calculate how much BTC at a certain rate he needs to refund, then go to the wallet and refund it. 

For me personally, this process takes 10-15min.

**Describe the solution you'd like**
Have a refund button in the invoice, which opens a pop-up, where merchant enters the amount in fiat/btc selects the rate from the drop down (current rate, invoice rate or custom rate) and customers refund address. When the refund pop-up is filled, proceed to wallet button auto-fills the wallet where a refund can be issued.

It would be nice if we'd have a refund field in the invoice itself that would auto-populate if the refund is issued.

**Provide examples**
![Screen Shot 2019-06-15 at 10 41 28](https://user-images.githubusercontent.com/36959754/59549198-404b0e00-8f5a-11e9-99a0-312038941611.png)

**Comment by woutersamaey** [comment]
Can we also have this as an API call? This seams like a more advanced eCommerce case that I’d drive from the online store software (Magento in my case).

**Comment by pavlenex** [comment]
@woutersamaey Sure, but the merchant would use their BTCPay internal wallet to send the funds, then we'll need API calls for that, but not sure how you can fit all of that into Magento's UI. To refund merchant would need to go to BTCPay find an invoice, sign it with a wallet and broadcast.

**Comment by woutersamaey** [comment]
@pavlenex it would make sense to have a refund API call, which queues the refund for later signing.
In a larger organisation, a supervisor could sign all pending transactions at the end of the day while the API was used to queue these.

**Comment by pavlenex** [comment]
Yes that's a great example, I agree, was not thinking big on this one.

**Comment by Kidist-Abraham** [comment]
Is this feature added yet?

**Comment by pavlenex** [comment]
No, otherwise it would be closed.

**Comment by Kukks** [comment]
> It's just a bit tricky to handle it as the flow is a bit interactive
* User contacts merchant requesting refund/merchant decides to cancel order and refund
* Merchant initiates refund process, decide amount to refund, select options of refund to presdent to user( if you have btc onchain and offchain, you may want to allow the buyer to choose based on preference, or else restrict to the one that benefits you best.)
* User provides payment info ( ugh, a reverse invoice kind fo thing, where BTCPay needs to ask for an address or a bolt11)
* Merchant approves and signs( if necessary, for example LN is not necesary to sign, and if you use the hot wallet feature in btcpay too)

**Comment by NicolasDorier** [comment]
This will be based on the new pull payments system backed into BTCPay Server the flow will be the following.

1. Customer pays
2. Customer contact the merchant to get a refund
3. The merchant accept, go to the invoice detail page
4. Click Refund
5. Set parameters relative to the refund (TO BE DEFINED @pavlenex )
6. A pull payment page get created, the merchant forward the link to the customer.
7. The customer browse to the page and pull the payment by entering his bitcoin address / shitcoin address / WebLN

In a nutshell, this replace the current flow 4 steps flow:

```
Customer:  I want refund
Merchant:  Ok, want is your address, do you want on chain or off chain, you will have to pay the fees is it ok?
Customer:  On chain bc1diwefiufw
Merchant:  Do the refund
```

To  a 2 steps flow:
```
Customer: I want refund
Merchant: Ok, here is the link, it is valid for 7 days.
```

**Comment by pavlenex** [comment]
How do you guys feel about this one?

![New Mockup 1](https://user-images.githubusercontent.com/36959754/80795326-74dcae00-8b9c-11ea-8de3-1af98f440a7a.png)

**Comment by Kukks** [comment]
https://support.bitpay.com/hc/en-us/articles/360000051746-How-to-claim-a-refund-from-a-BitPay-merchant-refund-email

**Comment by Kukks** [comment]
![image](https://user-images.githubusercontent.com/1818366/80795906-26c8aa00-8b9e-11ea-81ea-1494f2e6ff65.png)
![image](https://user-images.githubusercontent.com/1818366/80795914-2cbe8b00-8b9e-11ea-9278-7d7563aa93d5.png)
![image](https://user-images.githubusercontent.com/1818366/80795918-31833f00-8b9e-11ea-878d-f1e675ba80e7.png)
![image](https://user-images.githubusercontent.com/1818366/80795931-3a741080-8b9e-11ea-8dd9-51c435150317.png)
![image](https://user-images.githubusercontent.com/1818366/80795947-4069f180-8b9e-11ea-8b7a-a22b9ddb9302.png)
![image](https://user-images.githubusercontent.com/1818366/80795952-42cc4b80-8b9e-11ea-8a3f-797445d65f38.png)

**Comment by pavlenex** [comment]
Nice @Kukks that's something close to what I had in mind. We'd certainly need a `refunded` state of the invoice.

Can you show me what happens when you're on `partial refund screen` would like to see how they solved that.

**Comment by pavlenex** [comment]
@Kukks can we re-use camera QR from the wallet so that people can taking selfies? 😆 

```If your refund is the equivalent of 1000 USD or more, you will also need to complete the BitPay ID verification process in addition to the steps above. You will need to provide your legal name, country of residence, and birthdate. Then, you will take pictures of an approved document and take a selfie```

**Comment by Kukks** [comment]
Personally I would:
* NOT add a refund status. If you do that, you hide the previous status and have a hard time realizing why the refund was issued. 
* Refunds should be a separate entity. Each "refund" could be linked to multiple invoices ( especially in case of a payment request)
* We should allow multiple "refunds" per invoice, in the case of multiple order processing issues ( e.g: one item was out of stock, so partial refund. Then items were faulty or damaged, so partial refund of remaining, etc)
* We should invest time to make the Scheduled Transactions feature that was talked about in payjoin spec, so that we can batch refunds + use it for enhanced payjoins.

**Comment by pavlenex** [comment]
A `refund` status of some sort is needed. I may want to trigger certain action in WooCommerce after a refund is issued, but we can't do it with the current plugin anyway. A new entity makes sense and I agree we need to have a refund linked to multiple invoices.

**Comment by woutersamaey** [comment]
@pavlenex if you need a status in WooCommerce, you should arrange it there and not in BTCPay Server. Every platform will have different issues with this. Magento for example would not have a refund status. 

@Kukks version is 100% and top notch (judging as an eCommerce dev with 10 years experience).

**Comment by pavlenex** [comment]
I am not sure if we're taking about the same statuses @woutersamaey?
I can give you a two use-cases within btcpay:
- Sorting invoices by `refunded`
- Exporting invoices `refunded` (accounting) - very important

**Comment by Kukks** [comment]
I think a "refund" label would be great, but labels are on txs not invoices. Refunds should appear as another list under "Invoice summary" in the invoice list

**Comment by Kukks** [comment]
A "refunded' status on an invoice is problematic:
* What if you did a partial refund
* What if you did multiple refunds
* What if the refund was only to refund an overPaid amount? Would you not want this invoice to be marked as completed only instead?

**Comment by pavlenex** [comment]
I'm aware of how we use labels, I was just thinking if we can-reuse what we use in the wallet. It is important that people can export refunded invoices for accounting.

How about `partially-refunded` and `refunded` statuses? 

> Would you not want this invoice to be marked as completed only instead?

True. But accounting-wise it's a problem.

**Comment by NicolasDorier** [comment]
We can't change the status. This a a very tricky change that has lots of corner cases that just can't be explained this way.

What happen if the invoice, ie: what if the invoice goes from paid to refunded then the transaction get double spent? Should it go back to "new". There is tons of corner case like that.

**Comment by pavlenex** [comment]
`Refunded label` that can be applied next to status and then we'd have to modify invoice export. I feel like we're creating another "payment request" situation, where code-wise it makes sense, but for users it won't.

**Comment by Kukks** [comment]
> 
> 
> `Refunded label` that can be applied next to status and then we'd have to modify invoice export. I feel like we're creating another "payment request" situation, where code-wise it makes sense, but for users it doesn't.

Payment requests are useful, just limited because of lack of API to realistically use them

**Comment by pavlenex** [comment]
> Payment requests are useful, just limited because of lack of API to realistically use them

I have not said they are not useful, they way we presented them to the user as a new entity (makes sense code-wise) but it would make more sense if ti was within the apps.

**Comment by Kukks** [comment]
I think in export, we could just have a `refundedAmount` relative to paid amount

**Comment by pavlenex** [comment]
Makes sense. Thinking if perhaps labelling  and comments can be useful for invoices as well. `Refunded` label. Comment `Refunded because of wrong size`.

**Comment by Kukks** [comment]
> 
> 
> > Payment requests are useful, just limited because of lack of API to realistically use them
> 
> I have not said they are not useful, they way we presented them to the user as a new entity (makes sense code-wise) but it would make more sense if ti was within the apps. That's why I am glad we're having this discussion looking into it from multiple angles.

Yes, in hindsight we should have made individual payment requests as an app and added a type filter dropdown on Apps, since the code for Apps vs Payment requests is almost identical

**Comment by Kukks** [comment]
> 
> 
> Makes sense. Thinking if perhaps labelling and comments can be useful for invoices as well. `Refunded` label. Comment `Refunded because of wrong size`.

Yes, I think there should be a comment section when approving a request and creating the "refund request link"

**Comment by pavlenex** [comment]
Ok, so on the front-end I'd keep it as functional and simple as possible, for the API, perhaps @woutersamaey can provide some interesting use-cases.

`Amount` - full amount in BTC by default that can be edited.
`Currency` - default invoice currency that can be edited (we can show on the side USD denomination like we show for the wallet
`Rate` Current rate & rate on date invoice was paid (maybe someone can help formulate & simplify the wording here)
`e-mail` by default inputs buyer's email from invoice, can be edited
`note` used for label comment and customer's comment
`expires on`- perhaps date field is too much, but was thinking we can re-use payment requests UI here. Defaults to 3 days.

![Screenshot 2020-05-01 at 13 10 06](https://user-images.githubusercontent.com/36959754/80801683-6c40a380-8bad-11ea-8133-055767f281d2.png)

**Comment by woutersamaey** [comment]
From my experience using statuses has a lot of drawbacks. Basically because you can only have 1 status at a time. Some platforms use multiple statuses to bridge this, like a general status (vague like “new”, “processing” and “complete”) and separate substatuses for payment and shipment. 

This always implies a specific workflow, which we cannot impose. Imagine a payment, a partial refund, another payment and another partial refund all for 1 order. Yes, sometimes people have fat finger and do stupid things, or they change their mind and you need to correct any possible scenario. 

The best solution I can think of is to use amounts, like the totalPaid and totalRefunded. Any platform that needs a status, can compare the numbers and perhaps apply 1% margin if they really need to. 

Just like @Kukks mentions, there can be multiple payments and multiple refunds, so we should treat them the same way.

If you really need statuses, then go with “partial_refund” and “fully_refunded” if totalRefund > 0 and totalDue == 0. Otherwise the existing statuses “partial_paid” or “fully_paid” will be used. 

But i think just the amounts will be more flexible.

**Comment by NicolasDorier** [comment]
> expires on- perhaps date field is too much, but was thinking we can re-use payment requests UI here. Defaults to 3 days.

This is essential, you don't want to issue a refund only to be found that somebody take it a year later. The problem is even worse if you use the current rate, as it gives an option to the customer.

**Comment by pavlenex** [comment]
`current rate` vs `old rate` with a tooltip is good. I agree expiration is essential, I am just wondering if adding a date field makes sense or should we use "day" field.

> Editing the Currency (say changing USD to EUR) makes no sense to me.

This goes back to the [original problem](https://github.com/btcpayserver/btcpayserver/issues/890#issue-456515025)

> Use case: Order total is $100, however one item is defective and merchants needs to issue a refund. The price for the item is $25.

In reality you want to give people option to enter fiat amount where we refund etc it's not to change US to EUR, it's more BTC to fiat and vice-versa.

**Comment by pavlenex** [comment]
#1689
